server/aesdsocket.c

```diff
@@ -138,74 +307,59 @@ int main(int argc, char *argv[])
+            terminate_main_thread = true;
+            break;
+        }
         struct sockaddr_in *client_info = (struct sockaddr_in *)&client_addr;
-        inet_ntop(client_info->sin_family, &client_info->sin_addr, client_ip, sizeof(client_ip));
-
+        inet_ntop(client_info->sin_family, &client_info->sin_addr, current_thread_data->client_ip, sizeof(current_thread_data->client_ip));
         // Log accepted connection
-        syslog(LOG_INFO, "Accepted connection from %s", client_ip);
-
-        // open a new file
-        int aux_fd_write = open(FILE_LOG_PATH, O_CREAT | O_WRONLY | O_APPEND, 0666);
-        if (aux_fd_write == -1)
+        syslog(LOG_INFO, "Accepted connection from %s", current_thread_data->client_ip);
+        current_thread_data->is_thread_complete = false;
+        current_thread_data->thread_fd = new_fd;
+        pthread_t thread_id;
+        int thread_stat = pthread_create(&thread_id, NULL, (void *)thread_task, current_thread_data);
+        if (thread_stat == -1)
         {
-            syslog(LOG_ERR, "Failed to open a new file");
-            continue; // skip iteration
+            syslog(LOG_ERR, "Unable to create thread");
+            terminate_main_thread = true;
+            break;
         }
-        int buffer_size = 1024;
-        int total_bytes = 0;
-        char *temp_buffer = (char *)malloc(buffer_size);
-        int bytes_recv;
-        while ((bytes_recv = recv(new_fd, temp_buffer + total_bytes, buffer_size - total_bytes - 1, 0)) > 0)
+        current_thread_data->id = thread_id;
+        SLIST_INSERT_HEAD(&head, current_thread_data, next_ptr);
+        curr_ptr = SLIST_FIRST(&head);
+        while (curr_ptr != NULL)
         {
-            total_bytes += bytes_recv;
-            // make a valid string
-            temp_buffer[total_bytes] = '\0';
-            // check for a new line
-            if (strchr(temp_buffer, '\n') != NULL)
-                break;
-            // if not then we need to increase buffer_size
-            buffer_size += buffer_size;
-            char *new_buffer = realloc(temp_buffer, buffer_size);
-            if (new_buffer == NULL)
+            itr_ptr = SLIST_NEXT(curr_ptr, next_ptr); // Store the next item before potential removal
+            if (curr_ptr->is_thread_complete)
             {
-                syslog(LOG_ERR, "Failed to extend buffer size!");
-                free(temp_buffer);
-                close(aux_fd_write);
-                close(new_fd);
-                return FAILURE;
+                // Join the completed thread
+                if (pthread_join(curr_ptr->id, NULL) != 0)
+                {
+                    syslog(LOG_ERR, "Unable to join thread: %s", strerror(errno));
+                }
+                else
+                {
+                    syslog(LOG_INFO, "Thread joined successfully");
+                }
+
+                // Remove from the list
+                SLIST_REMOVE(&head, curr_ptr, thread_node, next_ptr);
```

Comment:
I don't think it's safe to SLIST_REMOVE and then access the pointer from SLIST_NEXT.  Check out SLIST_FOREACH_SAFE